src/v2/controllers/vreplicaset_controller/trusted/rely_guarantee.rs

```diff
@@ -0,0 +1,167 @@
+                &&& etcd_obj.metadata.resource_version == req.obj.metadata.resource_version
+                &&& etcd_obj.metadata.owner_references.is_Some()
+                &&& exists |vrs: VReplicaSetView| 
+                    #[trigger] owner_references.contains(vrs.controller_owner_ref())
+            }
+    }
+}
+
+// All delete requests to pods carry a precondition on the resource version,
+// and other controllers don't try to delete pods owned by a VReplicaSet.
+pub open spec fn vrs_rely_delete_req(req: DeleteRequest) -> StatePred<ClusterState> {
+    |s: ClusterState| {
+        req.key.kind == Kind::PodKind ==>
+            req.preconditions.is_Some()
+            && req.preconditions.get_Some_0().resource_version.is_Some()
+            && !{
+                let etcd_obj = s.resources()[req.key];
+                let owner_references = etcd_obj.metadata.owner_references.get_Some_0();
+                &&& s.resources().contains_key(req.key)
+                &&& etcd_obj.metadata.resource_version.is_Some()
+                &&& etcd_obj.metadata.resource_version
+                    == req.preconditions.get_Some_0().resource_version
+                &&& etcd_obj.metadata.owner_references.is_Some()
+                &&& exists |vrs: VReplicaSetView| 
+                    #[trigger] owner_references.contains(vrs.controller_owner_ref())
+            }
+    }
+}
+
+pub open spec fn vrs_rely(other_id: int) -> StatePred<ClusterState> {
+    |s: ClusterState| {
+        forall |msg| {
+            &&& #[trigger] s.in_flight().contains(msg)
+            &&& msg.content.is_APIRequest()
+            &&& msg.src == HostId::Controller(other_id)
+        } ==> match msg.content.get_APIRequest_0() {
+            APIRequest::CreateRequest(req) => vrs_rely_create_req(req)(s),
+            APIRequest::UpdateRequest(req) => vrs_rely_update_req(req)(s),
+            APIRequest::UpdateStatusRequest(req) => vrs_rely_update_status_req(req)(s),
+            APIRequest::DeleteRequest(req) => vrs_rely_delete_req(req)(s),
+            _ => true,
+        }
+    }
+}
+
+// VRS Guarantee Condition
+
+// VRS only creates pods owned by a VReplicaSet.
+pub open spec fn vrs_guarantee_create_req(req: CreateRequest) -> StatePred<ClusterHistory> {
+    |h: ClusterHistory| {
+        let owner_references = req.obj.metadata.owner_references.get_Some_0();
+        &&& req.obj.kind == Kind::PodKind
+        &&& req.obj.metadata.owner_references.is_Some()
+        &&& exists |vrs: VReplicaSetView| 
+            #[trigger] owner_references.contains(vrs.controller_owner_ref())
+    }
+}
+
+// VRS only sends delete requests to pods known to have been in etcd and 
+// owned by a VReplicaSet at some point in history. The request will
+// carry a precondition that matches the resource version of the historical
+// etcd object.
+pub open spec fn vrs_guarantee_delete_req(req: DeleteRequest) -> StatePred<ClusterHistory> {
```

Comment:
The delete guarantee should not talk about the history. It should just say for any pod delete request sent by this vrs controller, if that object in etcd matches the RV in precondition, then that object is owned by VRS.

Reply:
You will have intermediate lemmas that talk about history, but they are only used for proof. They should not be part of the interface.